scheduler/cloudformation/cloudformation.go

```diff
@@ -1,5 +1,408 @@
+	ListStackResourcesPages(*cloudformation.ListStackResourcesInput, func(*cloudformation.ListStackResourcesOutput, bool) bool) error
+	DescribeStackResource(*cloudformation.DescribeStackResourceInput) (*cloudformation.DescribeStackResourceOutput, error)
+	DescribeStacks(*cloudformation.DescribeStacksInput) (*cloudformation.DescribeStacksOutput, error)
+	WaitUntilStackCreateComplete(*cloudformation.DescribeStacksInput) error
+	WaitUntilStackUpdateComplete(*cloudformation.DescribeStacksInput) error
+}
+
+type ecsClient interface {
+	ListTasksPages(input *ecs.ListTasksInput, fn func(p *ecs.ListTasksOutput, lastPage bool) (shouldContinue bool)) error
+	DescribeTaskDefinition(*ecs.DescribeTaskDefinitionInput) (*ecs.DescribeTaskDefinitionOutput, error)
+	DescribeTasks(*ecs.DescribeTasksInput) (*ecs.DescribeTasksOutput, error)
+	StopTask(*ecs.StopTaskInput) (*ecs.StopTaskOutput, error)
+	UpdateService(*ecs.UpdateServiceInput) (*ecs.UpdateServiceOutput, error)
+}
+
+// Template represents something that can generate a stack body. Conveniently
+// the same interface as text/template.Template.
+type Template interface {
+	Execute(wr io.Writer, data interface{}) error
+}
+
+// Templates should add metadata to the ECS resources with the following
+// structure.
+type serviceMetadata struct {
+	// This is the name of the process that this ECS service is for.
+	Name string `json:"name"`
+}
+
+// Scheduler implements the scheduler.Scheduler interface using CloudFormation
+// to provision resources.
+type Scheduler struct {
+	// Template is a text/template that will be executed using the
+	// twelvefactor.Manifest as data. This template should return a valid
+	// CloudFormation JSON manifest.
+	Template Template
+
+	// The ECS cluster to run tasks in.
+	Cluster string
+
+	// If true, wait for stack updates and creates to complete.
+	Wait bool
+
+	// stackName returns the name of the stack for the app.
+	stackName func(app string) string
+
+	// CloudFormation client for creating stacks.
+	cloudformation cloudformationClient
+
+	// ECS client for performing ECS API calls.
+	ecs ecsClient
+}
+
+// NewScheduler returns a new Scheduler instance.
+func NewScheduler(config client.ConfigProvider) *Scheduler {
+	return &Scheduler{
+		cloudformation: cloudformation.New(config),
+		ecs:            ecs.New(config),
+		stackName:      stackName,
+	}
+}
+
+// Submit creates (or updates) the CloudFormation stack for the app.
+func (s *Scheduler) Submit(ctx context.Context, app *scheduler.App) error {
+	stackName := s.stackName(app.ID)
+
+	buf := new(bytes.Buffer)
+	if err := s.Template.Execute(buf, app); err != nil {
+		return err
+	}
+
+	desc, err := s.cloudformation.DescribeStacks(&cloudformation.DescribeStacksInput{
+		StackName: aws.String(stackName),
+	})
+
+	if err, ok := err.(awserr.Error); ok && err.Message() == fmt.Sprintf("Stack with id %s does not exist", stackName) {
+		if _, err := s.cloudformation.CreateStack(&cloudformation.CreateStackInput{
```

Comment:
Although it may not be necessary now. For example, when using this backend, when I add a domain to an app and deploy, cloudformation will provision a new ELB and ecs services, then swap DNS over to the new ELB.